docs/posts/lms-for-data/index.qmd

```diff
@@ -0,0 +1,695 @@
+2. Define a scalar Python UDF to apply the Marvin function to a column
+3. Apply the Marvin function within the UDF
+
+And apply that UDF to our penguins table:
+
+```{python}
+t = (
+    t.mutate(sentiment=sentiment_analysis(t.poem))  # <1>
+    .relocate(t.columns[:3], "sentiment")  # <2>
+    .cache()  # <3>
+)
+t
+```
+
+1. Apply the UDF by mutating the table with the function
+2. Rearrange the columns we care about to the front
+3. Cache the table to avoid re-running the UDF
+
+### Entity extraction
+
+While not exactly named entity recognition, we can extract arbitrary entities
+from text. In this case, we'll extract a list of words that rhyme from the poem:
+
+```{python}
+@ibis.udf.scalar.python  # <1>
+def extract_rhyming_words(text: str) -> list[str]:  # <1>
+    words = marvin.extract(  # <2>
+        text,  # <2>
+        instructions="Extract the primary rhyming words from the text",  # <2>
+    )  # <2>
+
+    return words  # <3>
+```
+
+1. Define a scalar Python UDF to extract rhyming words from a poem
+2. Use Marvin to extract the rhyming words from the poem
+3. Return the list of extracted words
+
+And apply that UDF to our penguins table:
+
+```{python}
+t = (
+    t.mutate(rhyming_words=extract_rhyming_words(t.poem))  # <1>
+    .relocate(t.columns[:4], "rhyming_words")  # <2>
+    .cache()  # <3>
+)
+t
+```
+
+1. Apply the UDF by mutating the table with the function
+2. Rearrange the columns we care about to the front
+3. Cache the table to avoid re-running the UDF
+
+### Translation
+
+We can translate the penguin poems into Spanish or any language the language
+model sufficiently knows:
+
+```{python}
+@marvin.fn  # <1>
+def _translate_text(text: str, target_language: str = "spanish") -> str:  # <1>
+    """Translate `text` to `target_language`."""  # <1>
+
+
+@ibis.udf.scalar.python  # <2>
+def translate_text(text: str, target_language: str = "spanish") -> str:  # <2>
+    return _translate_text(text, target_language)  # <3>
+```
+
+1. Define a Marvin function to translate text
+2. Define a scalar Python UDF to apply the Marvin function to a column
+3. Apply the Marvin function within the UDF
+
+And apply that UDF to our penguins table:
+
+```{python}
+t = (
+    t.mutate(translated_poem=translate_text(t.poem))  # <1>
+    .relocate(t.columns[:5], "translated_poem")  # <2>
+    .cache()  # <3>
+)
+t
+```
+
+1. Apply the UDF by mutating the table with the function
+2. Rearrange the columns we care about to the front
+3. Cache the table to avoid re-running the UDF
+
+## Writing analytical code
+
+Finally, language models can be used to write code. This is more useful with
+systems around them to execute code, feed back error messages, make adjustments,
+and so on. There are numerous pitfalls with language models writing code, but
+they're fairly good at SQL.
+
+### Three approaches
+
+We can think of three approaches to analytical code with language models:
+
+1. Use LMs in an analytic subroutine
```

Comment:
like an agent?

Reply:
no, just what's shown above -- basically using LLMs in UDFs